src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSource.cs

```diff
@@ -861,6 +886,55 @@ private void OnSuggestedActionsChanged(Workspace currentWorkspace, DocumentId cu
 
                 Volatile.Write(ref _lastSolutionVersionReported, solutionVersion);
             }
+
+            public async Task<ISuggestedActionCategorySet> GetSuggestedActionCategoriesAsync(ISuggestedActionCategorySet requestedActionCategories, SnapshotSpan range, CancellationToken cancellationToken)
+            {
+                var provider = _owner;
+                using (var asyncToken = _owner.OperationListener.BeginAsyncOperation(nameof(GetSuggestedActionCategoriesAsync)))
+                {
+                    var selection = await GetSpanAsync(range).ConfigureAwait(false);
+                    if (selection != null)
+                    {
+                        var document = range.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+                        using (var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken))
+                        {
+                            var linkedToken = linkedTokenSource.Token;
+
+                            var errorTask = Task.Run(
+                                () => GetFixLevelAsync(provider, document, range, linkedToken));
+
+                            var refactoringTask = Task.Run(
+                                () => TryGetRefactoringSuggestedActionCategoryAsync(provider, document, selection, linkedToken));
+
+                            var firstCompletedTask = await Task.WhenAny(errorTask, refactoringTask).ConfigureAwait(false);
+                            if (firstCompletedTask == errorTask)
```

Comment:
man... this pattern looks so close to what you wrote higher up.  I wonder if we can extract out a pattern.